docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,83 @@
+# Device Service Filters
+
+## Status
+
+### Proposed
+
+- design for Hanoi
+- implementation for Ireland
+
+## Context
+
+In EdgeX today, sensor/device data collected can be "filtered" by [application services](../../../microservices/application/ApplicationServices.md) before being exported or sent to some [north side](../../../general/Definitions.md#south-and-north-side) application or system. Built-in application service functions (available through the app services SDK) allow EdgeX event/reading objects to be filtered by device name or by value descriptor type.  That is, event/readings can be filtered by:
+
+- which device sent the event/reading (as determined by the device name in the event/readings)
+- the category (such as temperature or humidity) of data as determined by the value descriptor.
+
+*With value descriptor's deprecation, what will the app function SDK offer for filtering?*
+
+### Filter by device name and value descriptor nearer collection
+
+While application service filters help to limit the amount of data sent north, those events and readings are still collected, stored, processed by the EdgeX micro services, potentially resulting in wasted resources processing unneeded data.
+
+Therefore, there is a desire to help filter data at the device service level - thereby limiting the amount of sensor data that is captured, stored and processed as well as sent north.
+
+In the first stage of device service filtering, it is desired that the filter by device name and value descriptor be offered in all device services (via the SDK).  The filtering of event/readings should be accomplished just before the event/readings are sent to core data or (or put on the message bus in future implementations where DS message application services).  See Design below for more details.

```

Comment:
DeviceName and ValueDescriptorName filter doesn't make sense to me at the DS level. If filtering for these, is no different than now, so then one would filter them out, Which essentially would disable a Device or Resource.  Is filtering the appropriate way to accomplish these and does this add value? If I added a device, why would I then want to filter it out?

To me the value of filtering at the DS level would be elimination of duplicate values or values out of a certain range.

Reply:
Device services can represent more than one device.  I could see circumstances where some device readings need to be temporarily weeded out.  As things become more sophisticated, weed out under multiple condition filters (ex:  for deviceXYZ, filter out values outside range A-B.  For deviceABC, filter out value outside of range M-N).

As for value descriptor - what if a sensor, like a thermostat, was always sending in temp and humidity but you really only wanted the temp.  Why wait all the way until the app service to deal with this kind of filtering.  Save storage and possibly network transportation of the information that is not wanted.

Also - taking a crawl, walk, run approach, my thought was we tackle what we do already as a first step.  As we develop more capability and determine how to do this in a way that allows better/more sharing between app service functions, we can have an array of functions that could be applied in either service level - and then let the users decide when and which to use where.  It is just a library of filter functions to be applied as needed.

Reply:
Tony - I thought a bit about this, and I was concerned about being able to use (or rather reuse) what App Services has done - which means dealing with it at the Event/Reading level and not Device resource level.  It is a matter of whether we want to think about sharing this kind of functionality.